### PR TITLE
feat(func/greatest-least): new config param ignore null for func greatest/least

### PR DESCRIPTION
**Pull request by stephenkgu**
# Description

# Issue(s)

- Close/close/Fix/fix/Resolve/resolve: Issue Link
https://project.feishu.cn/taosdata_td/feature/detail/6511294180
# Checklist

Please check the items in the checklist if applicable.

- [ ] Is the user manual updated?
- [ ] Are the test cases passed and automated?
- [ ] Is there no significant decrease in test coverage?

### CODE REVIEW

File: source/libs/scalar/src/sclfunc.c [hunk 1]
```diff
@@ -6181,45 +6189,97 @@ static int32_t greatestLeastImpl(SScalarParam *pInput, int32_t inputNum, SScalar
...
-  pCovertParams = taosMemoryMalloc(inputNum * sizeof(SCovertScarlarParam));
-  for (int32_t j = 0; j < inputNum; j++) {
+
+  if (!ignoreNull) {
+    effectiveNum = dataInputNum;
+  }
+  pCovertParams = taosMemoryCalloc(effectiveNum, sizeof(SCovertScarlarParam));
+  if (pCovertParams == NULL) {
+    SCL_ERR_JRET(terrno);
+  }
```
[the remark this answers is not in the document]
Comment (stephenkgu) [reply source/libs/scalar/src/sclfunc.c:6473]: freeSCovertScarlarParams already guards against NULL

File: source/libs/scalar/src/sclfunc.c [hunk 2]
```diff
@@ -6176,50 +6184,93 @@ static int32_t greatestLeastImpl(SScalarParam *pInput, int32_t inputNum, SScalar
...
     }
+    outIdx++;
   }
 
   resultColIndex = taosMemoryCalloc(numOfRows, sizeof(int32_t));
-  SCL_ERR_JRET(vectorCompareAndSelect(pCovertParams, numOfRows, inputNum, resultColIndex, order));
+  if (resultColIndex == NULL) {
+    SCL_ERR_JRET(terrno);
+  }
+  SCL_ERR_JRET(vectorCompareAndSelect(pCovertParams, numOfRows, effectiveNum, resultColIndex, order, ignoreNull));
```
[the remark this answers is not in the document]
Comment (stephenkgu) [reply source/libs/scalar/src/sclfunc.c:6499]: The reviewer's premise is wrong: SCL_ERR_JRET already does goto _return

**Review by JinqingKuang — COMMENTED** [review]
## Code Review 总结

审查发现 **1 个 Critical** 和 **2 个 High** 级别问题，建议修复后合并。

### 统计
- Critical: 1 个（内存安全）
- High: 2 个（内存泄漏 + 可维护性）
- Medium: 2 个（文档 + 测试覆盖）

详细问题见下方评论。

**Comment by JinqingKuang** [comment]
## Code Review 总结

审查发现 **1 个 Critical** 和 **2 个 High** 级别问题，建议修复后合并。

### 统计
- 🔴 Critical: 1 个（内存安全）
- 🟠 High: 2 个（内存泄漏 + 可维护性）
- 🟡 Medium: 2 个（文档 + 测试覆盖）

---

### 问题 1: 🔴 Critical - 空指针解引用风险
**文件**: `source/libs/scalar/src/sclfunc.c:312-321`

`typeGetTypeModFromColInfo(&pInput[inputNum - 1].columnData->info)` 未检查 `info` 字段是否有效。如果结构体未正确初始化，可能触发段错误。

**修复建议**：
```c
if (pInput[inputNum - 1].columnData == NULL ||
    pInput[inputNum - 1].columnData->pData == NULL ||
    pInput[inputNum - 1].columnData->info.type == 0) {
  qError("greatestLeast: invalid flag column, func:%s", __FUNCTION__);
  code = TSDB_CODE_TSC_INTERNAL_ERROR;
  goto _return;
}
```

---

### 问题 2: 🟠 High - 内存泄漏风险
**文件**: `source/libs/scalar/src/sclfunc.c:378-424`

如果 `vectorConvertSingleCol` 失败导致循环提前退出，`outIdx` 会小于 `effectiveNum`。此时 `freeSCovertScarlarParams(pCovertParams, effectiveNum)` 会访问未初始化的内存槽位。

**修复建议**：
```c
_return:
  freeSCovertScarlarParams(pCovertParams, outIdx);  // 使用实际填充的数量
  taosMemoryFree(resultColIndex);
  return code;
```

需要在函数开头将 `outIdx` 初始化为 0，并在 `_return` 标签前确保其值正确。

---

### 问题 3: 🟠 High - 复杂嵌套逻辑
**文件**: `source/libs/scalar/src/sclfunc.c:270-299`

`vectorCompareAndSelect` 中的 NULL 处理逻辑嵌套过深（3 层嵌套 + 多个条件分支），圈复杂度高，难以维护和测试。

**建议重构**：
```c
static int32_t vectorCompareAndSelect(SCovertScarlarParam *pParams, int32_t numOfRows,
                                      int32_t numOfCols, int32_t *resultColIndex,
                                      EOperatorType optr, bool ignoreNull) {
  if (ignoreNull) {
    return vectorCompareAndSelectIgnoreNull(pParams, numOfRows, numOfCols, resultColIndex, optr);
  } else {
    return vectorCompareAndSelectDefault(pParams, numOfRows, numOfCols, resultColIndex, optr);
  }
}
```

将两种模式拆分为独立函数，降低单个函数的复杂度。

---

### 问题 4: 🟡 Medium - 文档不一致
**文件**: `docs/en/14-reference/01-components/01-taosd.md:9`

英文文档只有一行简短描述，而中文文档（`docs/zh/14-reference/01-components/01-taosd.md:68-77`）有完整的参数说明（类型、默认值、最小值、最大值、参数类型、动态修改、支持版本）。

**建议**：英文文档补充完整的参数说明格式，与中文文档保持一致的结构。

---

### 问题 5: 🟡 Medium - 测试覆盖不足
**文件**: `test/cases/11-Functions/01-Scalar/test_fun_sca_greatest_least.py`

当前测试缺少以下边界场景：
1. 空表 + 全 NULL 列的组合（`effectiveNum=0` 且 `numOfRows=0`）
2. 超大列数（50+ 列）的性能和正确性
3. 混合常量和列输入时的 broadcast 边界

**建议添加**：
- `case_empty_table_all_null_columns()`
- `case_many_columns_boundary()` (测试 50+ 列)
- `case_mixed_constant_column_broadcast()`

这些场景可以在后续 PR 中补充。

**Comment by stephenkgu** [comment]
1, info.type 在 translate 的时候赋值，这里不是指针，即便数据被踩乱了也不会 segfault
2, 内存申请是 calloc, 不会造成 memleak
3, 这里 cyclomatic complexity 约 6～8, 无需重构
4, 中英文档整体格式不同
5, Three new boundary cases added to TS & test suite.

**Review by JinqingKuang — APPROVED** [review]

**Review by guanshengliang — APPROVED** [review]
